### PR TITLE
[Diagnostics] Improvements for closure argument/result diagnostics

### PR DESCRIPTION
**Pull request by xedin**
Various improvements to the closure diagnostics.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@rudkx please take a look, this changes a way we do diagnostics for single argument trailing closures. This improves a bunch of situations related to standard methods like `map` etc. I'm still trying to figure out how to improve situations where trailing closures like this are used inside of sub-expressions too.

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
Improved couple more situations related to trailing closures.

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test

File: test/expr/closure/closures.swift [hunk 1]
```diff
@@ -14,7 +14,8 @@ var closure3a : () -> () -> (Int,Int) = {{ (4, 2) }} // multi-level closing.
 var closure3b : (Int,Int) -> (Int) -> (Int,Int) = {{ (4, 2) }} // expected-error{{contextual type for closure argument list expects 2 arguments, which cannot be implicitly ignored}}  {{52-52=_,_ in }}
 var closure4 : (Int,Int) -> Int = { $0 + $1 }
 var closure5 : (Double) -> Int = {
-       $0 + 1.0 // expected-error {{cannot convert value of type 'Double' to closure result type 'Int'}}
+       $0 + 1.0
+       // expected-error@+1 {{cannot convert value of type 'Double' to closure result type 'Int'}}
```
Comment (rudkx) [inline test/expr/closure/closures.swift]: Is there a way to get the diagnostic to point to the line that the expression is on? What happens if there is an explicit `return`?

File: test/Constraints/tuple_arguments.swift [hunk 2]
```diff
@@ -1526,3 +1526,13 @@ rdar32301091_2 { _ in }
 // expected-error@-1 {{contextual closure type '(Int, Int) -> ()' expects 2 arguments, but 1 was used in closure body}} {{19-19=,_ }}
 rdar32301091_2 { x in }
 // expected-error@-1 {{contextual closure type '(Int, Int) -> ()' expects 2 arguments, but 1 was used in closure body}} {{19-19=,<#arg#> }}
+
+struct SR_5199 {}
+extension Sequence where Iterator.Element == (key: String, value: String?) {
+  func f() -> [SR_5199] {
+    return self.map { (key, value) in
+        // expected-error@-1 {{closure tuple parameter '(key: String, value: Optional<String>)' does not support destructuring}} {23-35=(arg) -> SR_5199}} {{9-9=\n        let (key, value) = arg\n        return }}
```
Comment (rudkx) [inline test/Constraints/tuple_arguments.swift]: I suspect this needs an update after my change related to SE-0110.

File: test/Constraints/tuple_arguments.swift [hunk 3]
```diff
@@ -1526,3 +1526,13 @@ rdar32301091_2 { _ in }
...
+        SR_5199()
+    }
+  }
+}
```
Comment (rudkx) [inline test/Constraints/tuple_arguments.swift]: Newline!

File: lib/Sema/ConstraintSystem.h [hunk 4]
```diff
@@ -747,6 +747,10 @@ enum class ConstraintSystemFlags {
   /// Set if the client prefers fixits to be in the form of force unwrapping
   /// or optional chaining to return an optional.
   PreferForceUnwrapToOptional = 0x02,
+
+  /// If set, this is going to prevent constraint system from erasing all
+  /// discovered solutions expect the best one.
```
Comment (rudkx) [inline lib/Sema/ConstraintSystem.h]: typo: expect => except

**Review by rudkx — APPROVED** [review]
Other than the couple comments/questions, LGTM!

**Comment by xedin** [comment]
@swift-ci please smoke test

File: test/expr/closure/closures.swift (hunk 1 above)
Comment (xedin) [inline test/expr/closure/closures.swift]: Ah, this is actually used to indicate that the error is related to the closure itself when it has a single expression itself, it's always going to point at the end of it.

**Pull request merged**